docs/proposals/external-lb-source-ip-preservation.md

```diff
@@ -0,0 +1,226 @@
+- [Appendix](#appendix)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# Overview
+
+Kubernetes provides an external loadbalancer service type which creates a virtual external ip
+(in supported cloud provider environments) that can be used to load-balance traffic to
+the pods matching the service pod-selector.
+
+## Motivation
+
+The current implemention requires that the cloud loadbalancer balances traffic across all
+Kubernetes worker nodes, and this traffic is then equally distributed to all the backend
+pods for that service.
+Due to the DNAT required to redirect the traffic to its ultimate destination, the return
+path for each session MUST traverse the same node again. To ensure this, the node also
+performs a SNAT, replacing the source ip with its own.
+
+This causes the service endpoint to see the session as originating from a cluster local ip address.
+*The original external source IP is lost*
+
+This is not a satisfactory solution - the original external source IP MUST be preserved for a
+lot of applications and customer use-cases.
+
+# Alpha Design
+
+This section describes the proposed design for
+[alpha-level](../../docs/devel/api_changes.md#alpha-beta-and-stable-versions) support, although
+additional features are described in [future work](#future-work).
+
+## Overview
+
+The double hop must be prevented by programming the external load balancer to direct traffic
+only to nodes that have local pods for the service. This can be accomplished in two ways, either
+by API calls to add/delete nodes from the LB node pool or by adding health checking to the LB and
+failing/passing health checks depending on the presence of local pods.
+
+## Traffic Steering using LB programming
+
+This approach requires that the Cloud LB be reprogrammed to be in sync with endpoint presence.
+Whenever the first service endpoint is scheduled onto a node, the node is added to the LB pool.
+Whenever the last service endpoint is unhealthy on a node, the node needs to be removed from the LB pool.
+
+This is a slow operation, on the order of 30-60 seconds, and involves the Cloud Provider API path.
+If the API endpoint is temporarily unavailable, the datapath will be misprogrammed till the 
+reprogramming is successful and the API->datapath tables are updated by the cloud provider backend.
+
+*The HealthCheck approach is superior, LB reprogramming is not the preferred option here*
+
+## Traffic Steering using Health Checks
+
+This approach requires that all worker nodes in the cluster be programmed into the LB target pool.
+To steer traffic only onto nodes that have endpoints for the service, we program the LB to perform
+node healthchecks. The kube-proxy daemons running on each node will be responsible for responding
+to these healthcheck requests from the cloud provider LB healthchecker.
+kube-proxy already watches for Service and Endpoint changes, it will maintain an in-memory lookup
+table indicating the number of local endpoints for each service.
```

Comment:
@girishkalele: The danger with doing health checking directly from the LB is that you might "overreact" during overload.  Two examples:
- An intermittent failure affects a significant portion (eg: ~half) of your servers.  The LB considers them unhealthy, and then dumps all the resulting load on the few remaining servers.  The additional load then causes a few more of them to fail, dumping *more* load on the *fewer* remaining servers, etc.  You've turned what could have been a partial outage into a complete outage.
- When a server tries to automatically start up from a catastrophe like the above, the first server the LB considers "healthy" will immediately get the full cluster load dumped on it.  Assuming that drives it into the ground, then the next server to appear "healthy" will get the same treatment, etc guaranteeing that no server is able to stay up.  You've turned what could have been a recovery into an indefinite outage.

One possible automated solution to both of these is to be able to say "no fewer than N% should be considered unhealthy", or some similar policy.  That means that if there *are* fewer than this many healthy, then you assume you're in some catastrophic situation, and you (effectively) drop some queries at the LB rather than overload the remaining servers that *are* healthy.

Not a big deal, and I fully expect you to forget about this comment until after you suffer your first DoS outage ;)  My note was only that it is useful to have an intermediate point to insert arbitrary business logic between the real health check and what the LB does.